File: tests/language-support/javascript/src/storage.js
/**
 * @fileoverview Business logic for bookshelf storage operations.
 */

/**
 * Adds a book to a shelf.
 * @param {import('./models.js').Shelf} shelf - The target shelf
 * @param {import('./models.js').Book} book - The book to add
 * @throws {Error} If the shelf is at capacity
 */
export function addBook(shelf, book) {
  if (shelf.books.length >= shelf.capacity) {
    throw new Error(`Shelf "${shelf.name}" is at capacity (${shelf.capacity})`);
  }
  shelf.books.push(book);
}

/**
 * Removes a book from a shelf by ISBN.
 * @param {import('./models.js').Shelf} shelf - The target shelf
 * @param {string} isbn - ISBN of the book to remove
 * @throws {Error} If no book with that ISBN exists on the shelf
 */
export function removeBook(shelf, isbn) {
  const index = shelf.books.findIndex((b) => b.isbn === isbn);
  if (index === -1) {
    throw new Error(`Book with ISBN ${isbn} not found on shelf "${shelf.name}"`);
  }
  shelf.books.splice(index, 1);
}

/**
 * Finds books by author using case-insensitive substring matching.
 * @param {import('./models.js').Shelf} shelf - The shelf to search
 * @param {string} author - Author name or substring to search for
 * @returns {import('./models.js').Book[]} Books whose author matches
 */
export function findByAuthor(shelf, author) {
  const query = author.toLowerCase();
  return shelf.books.filter((b) => b.author.toLowerCase().includes(query));
}

/**
 * Finds books published within an inclusive year range.
 * @param {import('./models.js').Shelf} shelf - The shelf to search
 * @param {number} startYear - Start of the year range (inclusive)
 * @param {number} endYear - End of the year range (inclusive)
 * @returns {import('./models.js').Book[]} Books within the year range
 */
export function findByYearRange(shelf, startYear, endYear) {
  return shelf.books.filter((b) => b.year >= startYear && b.year <= endYear);
}

/**
 * Returns a sorted copy of all books on the shelf, ordered by title.
 * Does not mutate the original shelf.
 * @param {import('./models.js').Shelf} shelf - The shelf to sort
 * @returns {import('./models.js').Book[]} Sorted copy of the books array
 */
export function sortByTitle(shelf) {
  return [...shelf.books].sort((a, b) => a.title.localeCompare(b.title));
}

/**
 * Checks whether the shelf has reached its maximum capacity.
 * @param {import('./models.js').Shelf} shelf - The shelf to check
 * @returns {boolean} True if the shelf is full
 */
export function isFull(shelf) {
  return shelf.books.length >= shelf.capacity;
}

/**
 * Generates a multi-line formatted report for the shelf.
 * @param {import('./models.js').Shelf} shelf - The shelf to report on
 * @returns {string} Formatted library report
 */
export function generateReport(shelf) {
  const total = shelf.books.length;
  const available = shelf.books.filter((b) => b.available).length;
  const availablePct = total > 0 ? Math.round((available / total) * 100) : 0;
  const capacityPct = Math.round((total / shelf.capacity) * 100);

  const authorCounts = {};
  for (const book of shelf.books) {
    authorCounts[book.author] = (authorCounts[book.author] ?? 0) + 1;
  }
  const authorLines = Object.entries(authorCounts)
    .map(([name, count]) => `  - ${name} (${count} books)`)
    .join('\n');

  const years = shelf.books.map((b) => b.year);
  const minYear = Math.min(...years);
  const maxYear = Math.max(...years);

  const bookLines = shelf.books
    .map((b) => {
      const marker = b.available ? '[+]' : '[-]';
      return `  ${marker} ${b.title} by ${b.author} (${b.year}) - ${b.isbn}`;
    })
    .join('\n');

  return [
    `=== Library Report: ${shelf.name} ===`,
    `Total books: ${total}`,
    `Available: ${available} / ${total} (${availablePct}%)`,
    `Capacity: ${total} / ${shelf.capacity} (${capacityPct}% full)`,
    ``,
    `Authors (${Object.keys(authorCounts).length} unique):`,
    authorLines,
    ``,
    `Year range: ${minYear} - ${maxYear}`,
    ``,
    `Books by availability:`,
    bookLines,
  ].join('\n');
}

File: tests/language-support/go/storage.go
package main

import (
	"fmt"
	"sort"
	"strings"
)

// AddBook adds a book to the shelf, returning an error if the shelf is full.
func AddBook(shelf *Shelf, book Book) error {
	if IsFull(shelf) {
		return fmt.Errorf("shelf %q is at capacity (%d)", shelf.Name, shelf.Capacity)
	}
	shelf.Books = append(shelf.Books, book)
	return nil
}

// RemoveBook removes the book with the given ISBN, returning an error if not found.
func RemoveBook(shelf *Shelf, isbn string) error {
	for i, b := range shelf.Books {
		if b.Isbn == isbn {
			shelf.Books = append(shelf.Books[:i], shelf.Books[i+1:]...)
			return nil
		}
	}
	return fmt.Errorf("book with ISBN %q not found", isbn)
}

// FindByAuthor returns all books whose author contains the query (case-insensitive).
func FindByAuthor(shelf *Shelf, author string) []Book {
	query := strings.ToLower(author)
	var result []Book
	for _, b := range shelf.Books {
		if strings.Contains(strings.ToLower(b.Author), query) {
			result = append(result, b)
		}
	}
	return result
}

// FindByYearRange returns all books published within [start, end] inclusive.
func FindByYearRange(shelf *Shelf, start, end int) []Book {
	var result []Book
	for _, b := range shelf.Books {
		if b.Year >= start && b.Year <= end {
			result = append(result, b)
		}
	}
	return result
}

// SortByTitle returns a copy of the shelf's books sorted alphabetically by title.
func SortByTitle(shelf *Shelf) []Book {
	result := make([]Book, len(shelf.Books))
	copy(result, shelf.Books)
	sort.Slice(result, func(i, j int) bool {
		return result[i].Title < result[j].Title
	})
	return result
}

// IsFull reports whether the shelf has reached its capacity.
func IsFull(shelf *Shelf) bool {
	return len(shelf.Books) >= shelf.Capacity
}

// GenerateReport produces a multi-line formatted summary of the shelf.
func GenerateReport(shelf *Shelf) string {
	total, available := len(shelf.Books), 0
	for _, b := range shelf.Books {
		if b.Available {
			available++
		}
	}
	availPct, capacityPct := 0, 0
	if total > 0 {
		availPct = available * 100 / total
		capacityPct = total * 100 / shelf.Capacity
	}
	// Collect author counts preserving insertion order for display.
	authorOrder, authorCount := []string{}, map[string]int{}
	minYear, maxYear := shelf.Books[0].Year, shelf.Books[0].Year
	for _, b := range shelf.Books {
		if _, seen := authorCount[b.Author]; !seen {
			authorOrder = append(authorOrder, b.Author)
		}
		authorCount[b.Author]++
		if b.Year < minYear {
			minYear = b.Year
		}
		if b.Year > maxYear {
			maxYear = b.Year
		}
	}
	var sb strings.Builder
	fmt.Fprintf(&sb, "=== Library Report: %s ===\n", shelf.Name)
	fmt.Fprintf(&sb, "Total books: %d\n", total)
	fmt.Fprintf(&sb, "Available: %d / %d (%d%%)\n", available, total, availPct)
	fmt.Fprintf(&sb, "Capacity: %d / %d (%d%% full)\n", total, shelf.Capacity, capacityPct)
	fmt.Fprintf(&sb, "\nAuthors (%d unique):\n", len(authorOrder))
	for _, a := range authorOrder {
		fmt.Fprintf(&sb, "  - %s (%d books)\n", a, authorCount[a])
	}
	fmt.Fprintf(&sb, "\nYear range: %d - %d\n", minYear, maxYear)
	fmt.Fprintf(&sb, "\nBooks by availability:\n")
	for _, b := range shelf.Books {
		mark := "+"
		if !b.Available {
			mark = "-"
		}
		fmt.Fprintf(&sb, "  [%s] %s by %s (%d) - %s\n", mark, b.Title, b.Author, b.Year, b.Isbn)
	}
	return sb.String()
}
